src/memoized.rs

```diff
@@ -18,16 +18,56 @@ use std::fmt::Debug;
 use std::fmt::Display;
 use std::fmt::Write;
 use std::hash::Hash;
+use std::marker::PhantomData;
 
 /// Memoized queries store the result plus a list of the other queries
 /// that they invoked. This means we can avoid recomputing them when
 /// none of those inputs have changed.
-pub struct MemoizedStorage<DB, Q>
+pub type MemoizedStorage<DB, Q> = WeakMemoizedStorage<DB, Q, AlwaysMemoizeValue>;
+
+/// "Dependency" queries just track their dependencies and not the
+/// actual value (which they produce on demand). This lessens the
+/// storage requirements.
+pub type DependencyStorage<DB, Q> = WeakMemoizedStorage<DB, Q, NeverMemoizeValue>;
+
+pub struct WeakMemoizedStorage<DB, Q, M>
```

Comment:
I'd probably just call this `MemoizedStorage` and remove the type aliases, then adjust the macro — we don't really expect people to write these names by hand, do we? But these names / aliases are ok. Whichever you prefer I guess.

Reply:
I've been thinking about eventually merging `VolatileStorage` into this struct, such that it will probably want to be called `DerivedStorage` (vs `InputStorage`), with the "policy" determining other details.